File: routes/crop_calendar.py
from flask import Blueprint, render_template, request, jsonify

from services.crop_calendar_service import (
    generate_crop_calendar
)

# ==========================================
# Blueprint
# ==========================================

crop_calendar_bp = Blueprint(
    "crop_calendar",
    __name__
)

# ==========================================
# Crop Calendar Page
# ==========================================

@crop_calendar_bp.route("/crop_calendar")
def crop_calendar():

    return render_template("crop_calendar.html")


# ==========================================
# Generate AI Crop Calendar
# ==========================================

@crop_calendar_bp.route("/generate_crop_calendar", methods=["POST"])
def generate_calendar():

    try:

        data = request.get_json()

        crop = data.get("crop", "").strip()
        state = data.get("state", "").strip()

        if crop == "" or state == "":

            return jsonify({

                "error": "Please select both Crop and State."

            }), 400

        calendar = generate_crop_calendar(
            crop,
            state
        )

        return jsonify({

            "calendar": calendar

        })

    except Exception as e:

        print("Crop Calendar Route Error:", e)

        return jsonify({

            "error": "Unable to generate crop calendar."

        }), 500

File: static/js/crop_calendar.js
async function generateCalendar() {

    const cropValue = crop.value;
    const stateValue = state.value;

    if (cropValue === "" || stateValue === "") {
        alert("Please select Crop and State.");
        return;
    }

    loading.classList.remove("d-none");

    result.innerHTML = "";

    calendarBtn.disabled = true;
    calendarBtn.innerHTML = "Generating...";

    try {

        const response = await fetch("/generate_crop_calendar", {
            method: "POST",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify({
                crop: cropValue,
                state: stateValue
            })
        });

        const data = await response.json();

        loading.classList.add("d-none");

        calendarBtn.disabled = false;
        calendarBtn.innerHTML = "Generate Calendar";

        if (data.error) {
            result.innerHTML =
                `<div class="alert alert-danger">${data.error}</div>`;
            return;
        }

        const html = data.calendar
            .replace(/🌱/g, "<h4>🌱")
            .replace(/🌾/g, "<h4>🌾")
            .replace(/💧/g, "<h4>💧")
            .replace(/🧪/g, "<h4>🧪")
            .replace(/🌿/g, "<h4>🌿")
            .replace(/🐛/g, "<h4>🐛")
            .replace(/✂/g, "<h4>✂")
            .replace(/📦/g, "<h4>📦")
            .replace(/💰/g, "<h4>💰")
            .replace(/----------------------------------/g, "<hr>")
            .replace(/\n/g, "<br>");

        result.innerHTML =
            `<div class="calendar-card">${html}</div>`;

    }

    catch (error) {

        loading.classList.add("d-none");

        calendarBtn.disabled = false;
        calendarBtn.innerHTML = "Generate Calendar";

        result.innerHTML =
            `<div class="alert alert-danger">Something went wrong.</div>`;

        console.log(error);

    }

}

calendarBtn.addEventListener("click", generateCalendar);